ribbon.py

```diff
@@ -0,0 +1,545 @@
+
+    def approach(self, marker_image, ribbon_image, image):
+        aruco_id, aruco_x, aruco_y = self.check_for_aruco(image)
+        if aruco_id is not None:
+            logger.info ("Approaching turntable, aruco at %i, %i" % (aruco_x, aruco_y))
+            t_error  = (self.image_centre_x - aruco_x) / self.image_centre_x
+            dist_error = (aruco_y - self.APPROACH_DIST) / self.image_centre_y
+            if dist_error < self.APPROACH_TOL:
+                self.stage_number = 1
+                logger.info("Aruco triggered: At the turntable!")
+                self.drive.move(0, 0)
+            else:
+                ribbon_x, ribbon_y, ribbon_area, ribbon_contour = find_largest_contour(ribbon_image)
+                ribbon = [ribbon_x, ribbon_y, ribbon_area, ribbon_contour]
+                speed = max(min(self.SPEED_P * dist_error, self.MARKER_SPEED), -self.MARKER_SPEED)
+                self.follow_ribbon(ribbon, speed)
+        else:
+            #no auroc marker detected
+            ribbon_x, ribbon_y, ribbon_area, ribbon_contour = find_largest_contour(ribbon_image)
+            logger.info ("Approaching turntable, ribbon at %i, %i" % (ribbon_x, ribbon_y))
+            if ribbon_y < self.RIBBON_APPROACH_DIST:
+                self.stage_number = 1
+                logger.info("Ribbon triggered: At the turntable!")
+                self.drive.move(0, 0)
+            else:
+                ribbon = [ribbon_x, ribbon_y, ribbon_area, ribbon_contour]
+                self.follow_ribbon(ribbon, self.MARKER_SPEED)
+
+
+    def entry(self, marker_image, ribbon_image, image):
+        cropped_image = image[self.CROP_HEIGHT:(self.CROP_HEIGHT+45), 0:320]
+        blur_image = cv2.medianBlur(cropped_image, 3)
+        blur_image = cv2.cvtColor(blur_image, cv2.COLOR_RGB2HSV)
+        default_colour_bounds = ((40, 0, 0), (180, 255, 255))
+        limits = self.colour_bounds.get(
+            self.ALT_MARKER, default_colour_bounds
+        )
+        marker_mask = threshold_image(blur_image, limits)
+        marker_x, marker_y, marker_area, marker_contour = find_largest_contour(marker_mask)
+        pygame.mouse.set_pos(marker_y+self.CROP_HEIGHT, 320 - marker_x)
+        screen = pygame.display.get_surface()
+        frame = pygame.surfarray.make_surface(cv2.flip(marker_mask, 1))
+        screen.blit(frame, (self.CROP_HEIGHT, 0))
+        pygame.display.update()
+        if (marker_area > self.MIN_MARKER_AREA) and (marker_x > self.image_centre_x):  #>=go when marker on elft, <=go when marker on right
+            logger.info("Turntable barrier spotted at %d" % (marker_x))
+            self.drive.move(0, 1)
+            time.sleep(0.4)
+            self.drive.move(0, 0)
+
+    def leaving(self):
+        pass
+
+    def block_pushing(self):
+        pass
+
+    def direction(self, marker, ribbon):
+        '''function to check which side of the ribbon the tape marks are, indicating direction'''
```

Comment:
Point of order: It returns a Boolean, which doesn't indicate direction unless you explain what True and False mean. Even then, it's a *very* confusing way of returning a direction indication.

Probs fine though :+1: